lib/web_ui/lib/src/engine/html/picture.dart

```diff
@@ -603,3 +630,44 @@ class PersistedPicture extends PersistedLeafSurface {
     }
   }
 }
+
+/// Given size of a rectangle and transform, computes pixel density
+/// (scale factor).
+double _computePixelDensity(Matrix4? transform, double width, double height) {
+  if (transform == null || transform.isIdentity()) {
+    return 1.0;
+  }
+  final Float32List m = transform.storage;
+  // Apply perspective transform.
+  final double topLeftX = m[12] * m[15];
+  final double topLeftY = m[13] * m[15];
+  double x = width;
+  double y = height;
+  final double xp = (m[0] * x) + (m[4] * y) + m[12];
+  final double yp = (m[1] * x) + (m[5] * y) + m[13];
+  final double wp = 1.0 / ((m[3] * x) + (m[7] * y) + m[15]);
+  double scaleX = ((xp * wp) - topLeftX).abs() / width;
+  double scaleY = ((yp * wp) - topLeftY).abs() / height;
+  double scale = math.min(scaleX, scaleY);
+  if (scale < kEpsilon || scale == 1) {
+    // Handle local paint bounds scaled to 0, typical when using
+    // transform animations.
+    return 1.0;
+  }
+  if (scale > 1) {
+    // Normalize scale to multiples of 2: 1x, 2x, 4x, 6x, 8x.
+    // This is to prevent frequent rescaling of canvas during animations.
+    //
+    // On a fullscreen high dpi device dpi*density*resolution will demand
+    // too much memory, so clamp at 4.
+    scale = math.min(4.0, ((scale / 2.0).ceil() * 2.0));
```

Comment:
Should we normalize before the previous `if` statement? This way if `scale` snaps to 1x we just return 1.

Reply:
I like the early exit. For many apps most of the time scale is just 1.0 unless you use boxfit.